daml-lf/transaction/src/test/scala/com/digitalasset/daml/lf/validation/ValidationSpec.scala

```diff
@@ -0,0 +1,649 @@
+  // are careful to limit the combinational explosion to just what is necessary to ensure
+  // every tweak under test is hit by at least one pre-tweaked node.
+  //
+  // The testcases are organised so failure is detected and reported for a named tweak.
+
+  private class Tweak[X](val run: X => List[X])
+
+  private object Tweak {
+    def single[X](f1: X => Option[X]): Tweak[X] = {
+      def f(x: X) = f1(x) match {
+        case None => Nil
+        case Some(y) => List(y)
+      }
+      new Tweak(f)
+    }
+    def apply[X](f: X => List[X]): Tweak[X] = {
+      new Tweak(f)
+    }
+  }
+
+  //--[types]--
+
+  private type Val = V[V.ContractId]
+  private type KWM = KeyWithMaintainers[Val]
+  private type OKWM = Option[KWM]
+  private type Exe = NodeExercises[NodeId, V.ContractId]
+  private type Node = GenNode[NodeId, V.ContractId]
+  private type VTX = VersionedTransaction[NodeId, V.ContractId]
+
+  //--[samples]--
+
+  private val samPackageId1 = PackageId.assertFromString("packageId1")
+  private val samModuleName1 = ModuleName.assertFromString("moduleName1")
+
+  private val samLocation1 = Location(samPackageId1, samModuleName1, "someString1", (1, 2), (3, 4))
+  private val samLocation2 =
+    Location(samPackageId1, samModuleName1, "someString2", (11, 22), (33, 44))
+
+  private val samOptLocation1: Option[Location] = None
+  private val samOptLocation2: Option[Location] = Some(samLocation1)
+
+  private val samBool1 = true
+  private val samBool2 = false
+
+  private val samText1 = "some text"
+
+  private val samContractId1 = V.ContractId.V1(crypto.Hash.hashPrivateKey("cid1"))
+  private val samContractId2 = V.ContractId.V1(crypto.Hash.hashPrivateKey("cid2"))
+
+  private val samTemplateId1 = Identifier.assertFromString("-samPkg-:SamModule:samName1")
+  private val samTemplateId2 = Identifier.assertFromString("-samPkg-:SamModule:samName2")
+
+  private val samChoiceName1 = ChoiceName.assertFromString("Choice1")
+  private val samChoiceName2 = ChoiceName.assertFromString("Choice2")
+
+  private val samParty1 = Party.assertFromString("party1")
+  private val samParty2 = Party.assertFromString("party2")
+  private val samPartyX = Party.assertFromString("partyX")
+
+  private val samParties1: Set[Party] = Set()
+  private val samParties2: Set[Party] = Set(samParty1)
+  private val samParties3: Set[Party] = Set(samParty2)
+  private val samParties4: Set[Party] = Set(samParty1, samParty2)
+
+  private val samValue1: Val = V.ValueUnit
+  private val samValue2: Val = V.ValueContractId(samContractId1)
+
+  private val samKWM1 = KeyWithMaintainers(samValue1, samParties1)
+  private val samKWM2 = KeyWithMaintainers(samValue1, samParties2)
+  private val samKWM3 = KeyWithMaintainers(samValue2, samParties1)
+
+  private val samVersion1: TransactionVersion = TransactionVersion.minVersion
+  private val samVersion2: TransactionVersion = TransactionVersion.maxVersion
+
+  private val someCreates: List[Node] =
```

Comment:
This kind of code should broadly use `Seq` instead of `List`. It would probably be profitably idiomatic to replace every `List` with `Seq` here.

The only legitimate reasons to use `List` directly are

1. you need to traverse or
2. need a functor/monad instance, or
3. you rely specifically on the structure of a list, i.e. cheap push/pop